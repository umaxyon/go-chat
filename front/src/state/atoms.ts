import { atom } from "recoil"

export type Member = {
    user: string
}

export type FeedRow = {
    id: string,
    MessageType: "comment" | "addMember" | "leaveMember" | "system",
    user: string,
    text: string,
    createdAt: any
}

export type LoginData = {
    user: string,
    token: string,
    disconnect?: () => void
}

export const membersState = atom<Member[]>({
    key: "members",
    default: [],
})

export const commentsState = atom<FeedRow[]>({
    key: "comments",
    default: [],
})

export const loginState = atom<LoginData>({
    key: "loginUser",
    default: undefined,
})