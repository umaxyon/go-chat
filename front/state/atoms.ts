import { atom } from "recoil"

export type Member = {
    user: string
}

export type Comment = {
    id: string,
    user: string,
    text: string,
    createdAt: any
}

export type LoginData = {
    user: string,
    disconnect?: () => void
}

export const membersState = atom<Member[]>({
    key: "members",
    default: [],
})

export const commentsState = atom<Comment[]>({
    key: "comments",
    default: [],
})

export const loginState = atom<LoginData>({
    key: "loginUser",
    default: undefined,
})