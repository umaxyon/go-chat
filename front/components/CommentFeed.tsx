import { useQuery } from "@apollo/client"
import { useCallback, useEffect, useRef } from "react"
import { useRecoilState, useSetRecoilState } from "recoil"
import { commentsState, loginState, membersState } from "../state/atoms"
import { MESSAGE_QUERY, SUBSCRIPTION } from "../utils/client"
import CommentPanel from "./CommentPanel"


type CommentFeedProps = {
    user: string
}

const CommentFeed: React.FC<CommentFeedProps> = ({ user }) => {
    const [ comments, setComments ] = useRecoilState(commentsState)
    const [ login, setLogin ] = useRecoilState(loginState)
    const setMembers = useSetRecoilState(membersState)
    const scrollBottomRef = useRef<HTMLDivElement>(null);
    const result = useQuery(MESSAGE_QUERY)

    const scrollEnd = useCallback(() => {
        scrollBottomRef!.current!.scrollIntoView({ block: 'end' })
    }, [scrollBottomRef])

    useEffect(() => {
        if (result && result.data) {
            setComments(result.data.messages)
            if (result.data.members.length > 0) {
                setMembers(result.data.members)
            }
            scrollEnd()
        }
    }, [result, setComments, setMembers, scrollEnd])

    useEffect(()=> {
        const disconnect = result.subscribeToMore({
            document: SUBSCRIPTION,
            variables: { user },
            updateQuery: (prev, { subscriptionData }) => {
                if (!subscriptionData.data) return prev;
                const resp = subscriptionData.data.subscribe;
                if (resp.message) {
                    const messages = prev.messages ? [...prev.messages, resp.message] : [resp.message]
                    setComments(messages)
                    return Object.assign({}, prev, { messages })
                }
                if (resp.user) {
                    const members = prev.members ? [resp.user, ...prev.members] : [resp.user]
                    if (!members.some(m => m.user === login.user)) members.push(login)
                    setMembers(members)
                    setTimeout(scrollEnd, 0)
                    return Object.assign({}, prev, { members })
                }
            }
        })
        setLogin({ ...login, disconnect })
    // eslint-disable-next-line react-hooks/exhaustive-deps
    },[])

    const feed = !comments ? <></> : comments.map(c => {
        return <CommentPanel key={c.id} comment={c} />
    })

    return (
        <div className="
            border border-gray-300 min-w-3/4 h-full rounded-tr p-1
            overflow-y-scroll overflow-x-hidden scrollDiv">
            <div className="flex flex-col justify-end gap-1">{feed}</div>
            <div ref={scrollBottomRef} />
        </div>
    )
}
export default CommentFeed