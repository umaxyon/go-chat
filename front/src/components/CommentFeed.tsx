import { useCallback, useRef } from "react"
import { useRecoilValue } from "recoil"
import { feedsState } from "../state/atoms"
import CommentPanel from "./CommentPanel"
import InitLoad from "./InitLoad"


type CommentFeedProps = {
}

const CommentFeed: React.FC<CommentFeedProps> = () => {
    const feeds = useRecoilValue(feedsState)
    const scrollBottomRef = useRef<HTMLDivElement>(null);

    const scrollEnd = useCallback(() => {
        scrollBottomRef!.current!.scrollIntoView({ block: 'end' })
    }, [scrollBottomRef])

    const feed = !feeds ? <></> : feeds.map(c => {
        return <CommentPanel key={c.id} feed={c} />
    })

    return (
        <>
        <InitLoad afterAddMember={() => setTimeout(scrollEnd, 0) } />
        <div className="
            border border-gray-300 min-w-3/4 h-full rounded-tr p-1
            overflow-y-scroll overflow-x-hidden scrollDiv">
            <div className="flex flex-col justify-end gap-1">{feed}</div>
            <div ref={scrollBottomRef} />
        </div>
        </>
    )
}
export default CommentFeed