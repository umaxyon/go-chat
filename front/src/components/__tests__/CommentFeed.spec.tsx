import CommentFeed from "@/components/CommentFeed"
import { cleanup, screen, fireEvent, render } from '@testing-library/react';
import { useRecoilValue } from "recoil";
import renderer from "react-test-renderer"
import { Comment } from "src/state/atoms";


afterEach(() => cleanup())
jest.mock("../InitLoad", () => "InitLoad")
jest.mock("recoil");
const setRecoilMock = (val: Comment[]) => {
    (useRecoilValue as jest.Mock).mockImplementation(() => val)
}


const TEST_COMMENTS: Comment[] = [
    { id:   "1", user: "test_user1", text: "あああああ", createdAt: "2022-07-09T22:04:00.2890199+09:00" },
    { id:   "2", user: "test_user2", text: "いいいいい", createdAt: "2022-07-09T22:05:00.2890199+09:00" },
    { id:   "3", user: "test_user3", text: "ううううう", createdAt: "2022-07-09T22:06:00.2890199+09:00" },
]
const sut = (<CommentFeed />)

describe('CommentFeedテスト', () => {
    it("snapshot test", () => {
        setRecoilMock(TEST_COMMENTS)
        const component = renderer.create(sut);
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})