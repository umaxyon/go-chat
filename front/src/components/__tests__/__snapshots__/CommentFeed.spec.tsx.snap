// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentFeedテスト snapshot test 1`] = `
Array [
  <InitLoad
    afterAddMember={[Function]}
  />,
  <div
    className=" border border-gray-300 min-w-3/4 h-full rounded-tr p-1 overflow-y-scroll overflow-x-hidden scrollDiv"
  >
    <div
      className="flex flex-col justify-end gap-1"
    >
      <div
        className="w-full pt-2 text-xs"
      >
        <div
          className="flex flex-row w-full justify-between pb-1"
        >
          <div
            className="inline-block py-1 px-2.5 leading-none text-center whitespace-nowrap align-baseline font-bold bg-gray-200 text-gray-700 rounded-full"
          >
            test_user1
          </div>
          <div
            className="text-slate-400"
          >
            2022/07/09 22:04
          </div>
        </div>
        <div
          className="ml-7 p-2 bg-green-200 text-sm balloon rounded-lg"
        >
          あああああ
        </div>
      </div>
      <div
        className="w-full pt-2 text-xs"
      >
        <div
          className="flex flex-row w-full justify-between pb-1"
        >
          <div
            className="inline-block py-1 px-2.5 leading-none text-center whitespace-nowrap align-baseline font-bold bg-gray-200 text-gray-700 rounded-full"
          >
            test_user2
          </div>
          <div
            className="text-slate-400"
          >
            2022/07/09 22:05
          </div>
        </div>
        <div
          className="ml-7 p-2 bg-green-200 text-sm balloon rounded-lg"
        >
          いいいいい
        </div>
      </div>
      <div
        className="w-full pt-2 text-xs"
      >
        <div
          className="flex flex-row w-full justify-between pb-1"
        >
          <div
            className="inline-block py-1 px-2.5 leading-none text-center whitespace-nowrap align-baseline font-bold bg-gray-200 text-gray-700 rounded-full"
          >
            test_user3
          </div>
          <div
            className="text-slate-400"
          >
            2022/07/09 22:06
          </div>
        </div>
        <div
          className="ml-7 p-2 bg-green-200 text-sm balloon rounded-lg"
        >
          ううううう
        </div>
      </div>
    </div>
    <div />
  </div>,
]
`;
